### DISCUSSION TITLE
Proposal: "isin" operator on arrays

### DISCUSSION THREAD

**Discussion by CPosingies** [dotnet/csharplang#5082] [Language Ideas]
Hello there,

my proposal is, I guess, pretty simple. Assume you are parsing some data, and you must treat certain pieces in the data in a different way. What you already can do is:

```
using System.Collections.Generic;

var specials = new List<byte>(new byte[] {0, 7, 66, 123});
foreach (var item in chunk)
{
    if (specials.Contains(item))
        // do something useful here
}
```

My proposal is to simplify this syntax so you can write

```
// no extra using directive needed

var specials = new byte[] {0, 7, 66, 123};
foreach (var item in chunk)
{
    if (item isin specials)
        // do something even more useful here!
}
```

It should of course work for value types and for reference types.

The main aspect of my proposal is exactly this: the value types, without all the ado about generic collections. The second aspect is part of a general discussion on focus while reading code. the `foreach` loop's focus is `item`. Today (see above, first snippet), the focus is forcibly moved to properties of the `specials` entity, while my proposal truly stays with the `tem`. It may sound like syntactical sugar, but as far as I am concerned, the main difference between assembler code and C# code is exactly this: syntactical sugar to free your brain from technology-specific stuff so you can focus on domain- and problem-specific thinking.

**Comment by tacosontitan** [comment]
You're not really gaining anything useful because you're only dropping 5 characters from the example. Your first logical evaluation is 23 characters long (inside of the parenthesis) and the second is 18. Even in more complex situations, `.Contains` is likely the way to go, is already common place, and effective for what you're asking.

Also, IMHO, it looks a bit dirty conjoined. I would separate it for clarity and consistency with existing keywords:

    if (item is in specials)

Furthermore, you can use just `in`:

    if (item in specials)

However, even reducing it that far, it still falls into the category of high effort, little reward.

**Reply by CPosingies** [reply]
I would of course, in terms of readability and fluent code, prefer your "is in" solution, but still be happy with "isin" because the two words version would, would break the "f*** whitespace" rule applying to clean curly languages.

The major point still exists: no need for collections and linq-ish stuff, although i am a big fan of linqs. Just not where it messes up code readability.

**Reply by CyrusNajmabadi** [reply]
> would break the "f*** whitespace" rule applying to clean curly languages.

We have no such rule.  And we stronlgy support multi word constructs.  like `yield return ...` `is not ...`, etc. etc.  Also,  i would see `isin` and read that as either `i sin` (noice), or just think of it as some sort of funky math operator like `atan`.  :)

If we did ever do that it would be multi-word.  Though, that said, i think we'd want to just shorten that to `in` as i don't see the `is` providing value here.

**Reply by alrz** [reply]
I wouldn't go so far as to consider `is not` to be a multi-word construct 😄

**Reply by tacosontitan** [reply]
> I wouldn't go so far as to consider `is not` to be a multi-word construct 😄

`if (myVar isnt null)` 🤣 sorry, couldn't resist lol

**Comment by HaloFour** [comment]
You can accomplish something like this today via an extension method:

```cs
using System.Collections.Generic;

public static class Extensions {
    public static bool In<T>(this T value, List<T> list) => list.Contains(value);
}

class C {
    void M(byte[] chunk) {
        var specials = new List<byte>(new byte[] {0, 7, 66, 123});
        foreach (var item in chunk) {
            if (item.In(specials)) {
                // do something useful here
            }
        }
    }
}
```

I could otherwise see this considered as a part of pattern matching as a form of `in` pattern, although currently the operands in pattern matching must be literals or constants.

**Reply by alrz** [reply]
think this could be better off as a user-defined pattern `p is { Prop: Has(v) }`, but yeah, this would only work with constants.

**Comment by CPosingies** [comment]
Ack. "in" would be perfect! Now, imagine:
```
var f = Math.Sin * 115; // Volts /// sin referrer must be func<double, double>, not var, no other implicit operations
var range = Math2.Integrate(f, 0, 2 * Math.Pi) /// there is no Math2 namespace doing real calculus
if (hugo in range) ...
```

**Reply by tacosontitan** [reply]
> if (hugo in range) ...

I actually created extension methods to do just this, assuming that `hugo` implements `IComparable` (types such as `byte`, `string`, `int`, `long`, `ulong`, etc) and that `range` is a collection of types that implement `IComparable` (such as `byte[]`, `int[]`, etc.):

    public static bool WithinRange(this IComparable value, IComparable min, IComparable max) => value.CompareTo(min) >= 0 && value.CompareTo(max) <= 0;
    public static bool WithinRange<T>(this IComparable value, T[] values) where T : IComparable => value.WithinRange(values.Min(), values.Max());

The use is really straight forward:

    int hugo = 3;
    int[] values = Math2.Integrate(f, 0, 2 * Math.Pi);
    if (hugo.WithinRange(values)) ...
    // Or even:
    if (hugo.WithinRange(0, 50)) ...

**Reply by CPosingies** [reply]
I like your approach very much, but as you know, my comment was merely a "now more on this way" thing. By the way, I strongly support each and every effort to add maths and more maths to Dotnet! <345

`var t = new Tensor(...)`

**Comment by koszeggy** [comment]
I actually created a bunch of [extension methods](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/63781d58b559c9d15ad9979603a3823928e7a5a3/KGySoft.CoreLibraries/CoreLibraries/_Extensions/ObjectExtensions.cs#L107) for this.

The simplest overload uses `params T[]` but there link for some more (`IEnumerable<T>`, `object[]` and even `Func<T>[]` if the elements of the set are expensive to evaluate):

```cs
public static bool In<T>(this T item, params T[]? set)
{
    int length;
    if (set == null || (length = set.Length) == 0)
        return false;

    // of course, this can be a single line with LINQ .Any if performance does not matter
    var comparer = EqualityComparer<T>.Default;
    for (int i = 0; i < length; i++)
    {
        if (comparer.Equals(item, set[i]))
            return true;
    }

    return false;
}
```

And it can be used like this:

```cs
if (item.In(specials))
```
or like this:
```cs
if (item.In(0, 7, 66, 123)
```

But in fact if the set consists of constants, starting with C# 9 you can also write
```cs
if (item is 0 or 7 or 66 or 123)
```
which translates to `||` expressions without any method call.

**Reply by CPosingies** [reply]
As I said, I know how I could implement such stuff. My proposal is not rooting in a lack of C#'s abilities but in a lack of readability, plus the need of all the Col.Gen. stuff.

**Comment by momvart** [comment]
I would love an infix boolean operator like `in` or `is in`. Especially, when you come from kotlin or python lack of this syntax is disappointing.
I know that comparing to python, which is dynamic-typed and easy-going, and kotlin, with sophisticated operator overriding support, is not fair.
But with the help of LINQ, it could be simply applied between `T` and `IEnumerable<T>`. However, for complete support, it may require a definition for a new operator which I think is a lot of work.

**Reply by CPosingies** [reply]
I shall not reply to this! I shall not reply to this! I shall not rep... Python is ugly! I shall not reply to this! I shall not reply to this! I shall not reply to this! I shall not reply to this!

=P

(Seriously, whitespace's purpose is to provide semantics, not syntax.)

The concept of Kotlin in very interesting, but please tell your guys that there is no more need to abbr sth lk "function" ;-)

**Reply by CyrusNajmabadi** [reply]
@CPosingies The post you were replying to was not suggesting that we be whitespace sensitive in the same manner as python.  Simply that we offer a similar operator that that language offers.  In that regard, the suggestion makes sense and fits fine into the design space that C# occupies.

**Comment by CPosingies** [comment]
While further thinking about my proposal I'd now prefer a simple "in" infix, because it could not only be used like

`if (a in {1, 2, 3}) ...`

but also like

`if ({a, b} in {1, 2, 3})`

If the powers to be prefer "is in" plus "are in": be it so. Even more readabily, but a new keyword. I'd be happy with either way.

**Reply by tacosontitan** [reply]
I *personally* believe that `in` alone still fits just fine. 😄

**Comment by DavidArno** [comment]
My take on this matter is that what's really being asked for is an `in` pattern:
```cs
if (item is in 0, 7, 66, 123) ...

item switch {
    in 0, 7, 66, 123 => ...
```

As others pointed out, the above example can be expressed as:
```cs
if (item is 0 or 7 or 66 or 123) ...

item switch {
    in 0 or 7 or 66 or 123 => ...
```
but there's a problem. I have a collection of magic numbers and magic numbers are rightly frowned upon. And this leads back to the OP where what was being asked for is effectively:
```cs
if (item is in specials) ...
```
But patterns require constants and it's not possible to declare a constant array. A number of people suggest extension methods, but then we don't have a pattern any more, so that's not a good solution. We could use a `when` guard:
```cs
item switch {
    int x when IsSpecialNumber(x) => ...
```
but that only works with switch expressions. We can't use a `when` guard with an `if`. But maybe that's not a problem as we can mix a pattern with other boolean expressions:
```cs
if (item is int x && IsSpecialNumber(x)) ...
```

Of course, what would be a really nice solution would be to declare `IsSpecialNumber` as an active pattern so it can be used in all patterns:
```cs
if (item is int x and SpecialNumber(x)) ...

item switch {
    int x and SpecialNumber(x) => ...
```
but active patterns remain wishful thinking only for C# as far as I know. One day maybe...

**Reply by tacosontitan** [reply]
All these years and we still can't declare arrays as constant. 🤕 

<sub>***Note**: We won't talk about pseudo-constant by using `readonly`.*</sub>

**Reply by CPosingies** [reply]
> ```cs
> if (item is in 0, 7, 66, 123) ...
> 
> item switch {
>     in 0, 7, 66, 123 => ...
> ```

> ```cs
> if (item is in specials) ...
> ```

What I am proposing would be a first step to treat (mathematical) sets as first class members in C#. Hints: ∈ ⊂ ⊆ ∃ ∀.

```cs
if (items ⊆ specials)...
```

(By the way:

```cs
if (|specials| > someMagicNumber) ...
```
since a set has no .Length and no .Count, but a cardinality.)

Just to mention this: a ```HashSet<T>``` is not a mathematical set because it cannot contain elements twice that are identical regarding their properties; while a mathematical set can: A set<string> for example { "Amber", "Brilliant", "Xavier", "Honorable", "Brilliant" }, and a random access would return "Brilliant" statistically twice as often as "Amber" or "Xavier".

Is this useful? Let's see...

```cs
var params = new Math.Set(Math.Sets.ParseString(args, new {" "}));
// start exe with -O0 -g -g -g
// params := {string -O0, string -g, string -g, string -g}
var debugRequests = params.Select(x => x == "-g");
var debugLevel = |debugRequests|;
if ("-O0" is in params) ...
```

**Reply by tacosontitan** [reply]
Absolute value seems like it could complicate some things 🤔 (from a compilation standpoint).

**Comment by jtleaming** [comment]
I was going to open a discussion, but this seems similar enough of a request. 

I've been wanting to replace something like this....
```cs
    var accepted = new[] { "one", "two", "three" };
```

```cs
    var ent = new Entity("John", 10);

    if (ent.Id > 0 && accepted.Contains(ent.Name))
    {
        Console.WriteLine("success");
    }
```
with this
```cs
    if (ent is {Id: > 0, Name: in accepted})
    {
        Console.WriteLine("success");
    }
```
I found this is doable
```cs
    if (ent is { Name: "one" or "two" or "three", Id: > 0 })
    {
        Console.WriteLine("success");
    }
```
but it feels less maintainable and harder to read.

I also think maybe `any` would be an acceptable key word here
`if(ent.Name is any acceptable)` of `if(ent is {Name: any acceptable})`

**Reply by FaustVX** [reply]
This would be possible with the Active pattern proposal #277
